LTO: wild: error: Undefined symbol WebPFree, referenced by

Noticed while building the `libwebp` Gentoo package, could be reproduced locally with:

```
wget https://storage.googleapis.com/downloads.webmproject.org/releases/webp/libwebp-1.6.0.tar.gz
tar xvzf libwebp-1.6.0.tar.gz
cd libwebp*
CFLAGS="-flto=auto -fuse-ld=wild" ./configure
make VERBOSE=1 V=1
...
make[1]: Entering directory '/tmp/libwebp-1.6.0/examples'
/bin/sh ../libtool  --tag=CC   --mode=link gcc -fvisibility=hidden -Wall -Wdeclaration-after-statement -Wextra -Wfloat-conversion -Wformat -Wformat-nonliteral -Wformat -Wformat-security -Wmissing-declarations -Wmissing-prototypes -Wold-style-definition -Wshadow -Wstrict-prototypes -Wundef -Wunreachable-code -Wunused-but-set-variable -Wunused -Wvla -flto=auto -fuse-ld=wild    -o dwebp dwebp-dwebp.o libexample_util.la ../imageio/libimagedec.la ../imageio/libimageenc.la ../imageio/libimageio_util.la ../src/libwebp.la -L/usr/lib -lpng16 -ljpeg
libtool: link: gcc -fvisibility=hidden -Wall -Wdeclaration-after-statement -Wextra -Wfloat-conversion -Wformat -Wformat-nonliteral -Wformat -Wformat-security -Wmissing-declarations -Wmissing-prototypes -Wold-style-definition -Wshadow -Wstrict-prototypes -Wundef -Wunreachable-code -Wunused-but-set-variable -Wunused -Wvla -flto=auto -fuse-ld=wild -o .libs/dwebp dwebp-dwebp.o  ./.libs/libexample_util.a ../imageio/.libs/libimagedec.a /tmp/libwebp-1.6.0/src/demux/.libs/libwebpdemux.so /tmp/libwebp-1.6.0/src/.libs/libwebp.so /tmp/libwebp-1.6.0/sharpyuv/.libs/libsharpyuv.so -lm ../imageio/.libs/libimageenc.a ../imageio/.libs/libimageio_util.a ../src/.libs/libwebp.so -L/usr/lib -lpng16 -ljpeg -Wl,-rpath -Wl,/usr/local/lib
wild: error: Undefined symbol WebPFree, referenced by
    /tmp/ccNZR24f.ltrans0.ltrans.o
collect2: error: ld returned 255 exit status
```

--- a/libwild/src/linker_plugins.rs
+++ b/libwild/src/linker_plugins.rs
@@ -537,7 +537,7 @@ impl<'data> LtoInput<'data> {
     }
 
     pub(crate) fn is_optional(&self) -> bool {
-        self.input_ref.has_archive_semantics()
+        self.input_ref.has_archive_semantics() && !self.input_ref.file.modifiers.whole_archive
     }
 }
 

--- a/wild/tests/sources/elf/exclude-libs-selective/exclude-libs-selective.c
+++ b/wild/tests/sources/elf/exclude-libs-selective/exclude-libs-selective.c
@@ -1,3 +1,4 @@
+//#Config:default
 //#LinkArgs:-z now -Bshareable --exclude-libs exclude-libs-selective-excluded.a
 //#Mode:dynamic
 //#RunEnabled:false
@@ -7,6 +8,11 @@
 //#DiffIgnore:.dynamic.DT_RELAENT
 //#ExpectDynSym:included_fn
 //#NoDynSym:excluded_fn
+
+// --whole-archive should not cause us to treat an archive like it's not an
+// archive for the purposes of --exclude-libs.
+//#Config:whole-archive:default
+//#LinkArgs:-z now -Bshareable --whole-archive --exclude-libs exclude-libs-selective-excluded.a
 
 extern int excluded_fn(void);
 extern int included_fn(void);

--- a/wild/tests/sources/elf/lto-whole-archive-shared/bar.c
+++ b/wild/tests/sources/elf/lto-whole-archive-shared/bar.c
@@ -1,0 +1,1 @@
+int bar(void) { return 42; }

--- a/wild/tests/sources/elf/lto-whole-archive-shared/lto-whole-archive-shared.c
+++ b/wild/tests/sources/elf/lto-whole-archive-shared/lto-whole-archive-shared.c
@@ -1,0 +1,12 @@
+//#RequiresLinkerPlugin:true
+//#Archive:bar.c
+//#LinkerDriver:gcc
+//#CompArgs:-flto
+//#LinkArgs:-Wl,-znow -flto -nostdlib -O0 -shared -Wl,--whole-archive
+//#DiffIgnore:section.got
+//#DiffIgnore:.dynamic.DT_RELA*
+//#RunEnabled:false
+//#ExpectDynSym:foo
+//#ExpectDynSym:bar
+
+int foo(void) { return 42; }

